File: Python/XorShift.py
from time import time
from copy import copy

class XorShift:
	defaults={
		"x":123456789,
		"y":362436069,
		"z":521288629,
		"w":88675123
	}

	def __init__(self,
		w=int(time()),
		x=None,
		y=None,
		z=None
	):
		if x is None: x=w<<13
		if y is None: y=(w>>9)^(x<<6)
		if z is None: z=y>>7
		self.seeds={"x":x,"y":y,"z":z,"w":w}
		self.randCount=0
		self.__r=self.randGen(w,x,y,z)

	def randGen(self,w,x,y,z):
		while True:
			t=x^(x<<11&0xFFFFFFFF)
			x,y,z=y,z,w
			w=(w^(w>>19))^(t^(t>>8))
			yield w

	def rand(self):
		self.randCount+=1
		return next(self.__r)

	def randInt(self,min=0,max=0x7FFFFFFF):
		return self.rand()%(max+1-min)+min

	def randFloat(self,min=0,max=1):
		return self.rand()%0xFFFF/0xFFFF*(max-min)+min

	def shuffle(self,_arr):
		arr=copy(_arr)
		if isinstance(arr,list):
			isList=True
		elif isinstance(arr,range) or isinstance(arr,tuple):
			isList=True
			arr=list(arr)
		else:
			isList=False
			arr=list(arr)

		for i in range(0,len(arr)-1):
			r=self.randInt(i,len(arr)-1)
			arr[i],arr[r]=arr[r],arr[i]

		if not isList:
			arr=iter(arr)
		return arr

	def defaultSeed():return type("defaultSeed",(XorShift,),{
		"__init__":lambda self:
			XorShift.__init__(
				self,
				**XorShift.defaults
			)
	})()

File: JavaScript/XorShift.js
"use strict";

module.exports=(()=>{
	const r=Symbol();
	class XorShift{
		constructor(
			w=0|Date.now(),
			x,y,z
		){
			if(x===undefined) x=0|w<<13;
			if(y===undefined) y=0|(w>>>9)^(x<<6);
			if(z===undefined) z=0|y>>>7;
			this.seeds={x:x>>>0,y:y>>>0,z:z>>>0,w:w>>>0};
			Object.defineProperty(this,"seeds",{writable:false});
			this.randCount=0;
			this[r]=this.randGen(w,x,y,z);
		}

		*randGen(w,x,y,z){
			var t;
			for(;;){
				t=x^(x<<11);
				x=y;
				y=z;
				z=w;
				yield w=((w^(w>>>19))^(t^(t>>>8)))>>>0;
			}
		}

		rand(){
			this.randCount=0|this.randCount+1;
			return this[r].next().value;
		}

		randInt(min=0,max=0x7FFFFFFF){
			return 0|this.rand()%(max+1-min)+min;
		}

		randFloat(min=0,max=1){
			return Math.fround(this.rand()%0xFFFF/0xFFFF)*(max-min)+min;
		}

		shuffle(_arr){
			var arr=_arr.concat();
			for(let i=0;i<=arr.length-2;i=0|i+1){
				let r=this.randInt(i,arr.length-1);
				let tmp=arr[i];
				arr[i]=arr[r];
				arr[r]=tmp;
			}
			return arr;
		}
	}
	XorShift.defaults={
		x:123456789,
		y:362436069,
		z:521288629,
		w:88675123
	};

	XorShift.defaultSeed=class XorShift_default extends XorShift{
		constructor(){
			super(
				XorShift.defaults["w"],
				XorShift.defaults["x"],
				XorShift.defaults["y"],
				XorShift.defaults["z"]
			);
		}
	}

	Object.freeze(XorShift);

	return XorShift;
})();